rfd/0033-desktop-access.md

```diff
@@ -0,0 +1,176 @@
+  canvas
+- captures user input, encodes it into desktop protocol messages and sends them
+  over the websocket
+
+TODO: wireframes?
+
+### Configuration
+
+There are no additional configuration options for Desktop Access in the web UI
+or the proxy. The Desktop Access client is part of the web UI and uses the same
+web port on the proxy for connecting a websocket.
+
+All relevant configuration to enable the feature will live in OS-specific
+backend services, described in their own RFDs (like RFD 34 for Windows).
+
+### Authorization
+
+Role definitions for RBAC have a few new fields for Desktop Access:
+- `${OS}_desktop_logins` - list of desktop login names allowed/denied for
+  desktop hosts with a given OS
+- `${OS}_desktop_labels` - list of labels to match the target desktop hosts
+  against
+- `desktop_clipboard` - option for allowing copy/paste to/from the desktop
+  host; default is `true`
+- `record_desktop_session` - option for enabling/disabling recording of
+  sessions to target desktop hosts; for very long and not security-sensitive
+  sessions, recording can be disabled to save on storage costs; default is
+  `true`
+
+Example:
+```yaml
+kind: role
+version: v3
+metadata:
+  name: windows_db_admin
+spec:
+  options:
+    desktop_clipboard: true
+    record_desktop_session: true
+  allow:
+    # Allow windows sessions as user DBAdmin in test and staging environments.
+    windows_desktop_logins: ['DBAdmin']
+    windows_desktop_labels:
+      'env': ['staging', 'test']
+  deny:
+    # Prevent logins to prod environment or as Administrator.
+    windows_desktop_logins: ['Administrator']
+    windows_desktop_labels:
+      'env': ['prod']
+```
+
+### Storage schema
+
+There are 2 new kinds of objects stored in Teleport: `${OS}DesktopService` and
+`${OS}Desktop`. For example `WindowsDesktopService` and `WindowsDesktop`. The
+former tracks registration of an `${OS}_desktop_service` instance of Teleport.
+The latter tracks an individual target desktop host for logging into.
+
+OS-specific backend services are responsible for creating the `${OS}Desktop`
+objects, including their RBAC labels.
+
+The schema for these objects is defined in their respective RFDs.
+
+### Session recording
+
+Video output in our desktop protocol uses bitmaps and not a standard video
+encoding format. We will record the video output data as-is, with timestamps
+added to each protocol message (or perhaps
+[APNG](https://en.wikipedia.org/wiki/APNG)).
+
+To playback session recordings, we will replay them in the same canvas-based JS
+code as the original session used live.
+
+#### Recording export
+
+Some users might want to export the recording for sharing. To allow playback
+outside of the Teleport web UI, the exported recording should use a standard
+video format like MP4 or WebM.
+
+To convert our storage format to the video format, we will use [wasm-based
+ffmpeg](https://github.com/ffmpegwasm/ffmpeg.wasm) in the browser. Conversion
```

Comment:
Any experience with how ffmpeg performs in the browser? I imagine an older workstation with a lot of browser tabs open might get bogged down trying to transcode video.

Reply:
I haven't benchmarked it, but it will definitely be slow and resource-intensive.
However, this will be a relatively rare process - most users can watch the replay on our web UI without exporting and transcoding the video.